BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,328 @@ def browse
+    api_response = generate_register_response.merge({
+      :newly_registered_userfiles      => registered,
+      :previously_registered_userfiles => already_registered
+    })
 
-    if num_skipped > 0
-      flash[:notice] += "Skipped #{num_skipped} files.\n"
+    respond_to do |format|
+      format.html { redirect_to :action => :browse }
+      format.json { render :json => api_response }
+      format.json { render :xml  => api_response }
     end
+  end
 
-    if newly_registered_userfiles.size > 0
-      clear_browse_provider_local_cache_file(@as_user, @provider) unless request.format.to_sym == :xml || request.format.to_sym == :json
-      flash[:notice] += "Registered #{newly_registered_userfiles.size} files.\n"
-      if @as_user != current_user
-        flash[:notice] += "Important note! Since you were browsing as user '#{@as_user.login}', the files were registered as belonging to that user instead of you!\n"
+  # Unregister a list of files from CBRAIN
+  def unregister
+    # Extract key parameters & make sure the provider is browsable
+    @provider = DataProvider.find_accessible_by_user(params[:id], current_user)
+    @as_user  = browse_as(params['as_user_id'])
+    unless @provider.is_browsable?(current_user)
+      flash[:error] = "You cannot unregister files from this provider."
+      respond_to do |format|
+        format.html { redirect_to :action => :index }
+        format.xml  { render :xml  => { :error => flash[:error] }, :status => :forbidden }
+        format.json { render :json => { :error => flash[:error] }, :status => :forbidden }
       end
-    elsif num_erased > 0
-      clear_browse_provider_local_cache_file(@as_user, @provider) unless request.format.to_sym == :xml || request.format.to_sym == :json
-      flash[:notice] += "Erased #{num_erased} files.\n"
-    elsif num_unregistered > 0
-      flash[:notice] += "Unregistered #{num_unregistered} files.\n"
-    else
-      flash[:notice] += "No files affected.\n"
+      return
     end
 
-    # Nothing else do to if no automatic operation required.
-    if (move_or_copy != "MOVE" && move_or_copy != "COPY") || !new_dp || newly_registered_userfiles.size == 0
-      if newly_registered_userfiles.size > 0
-        CBRAIN.spawn_with_active_records(:admin, "Set Sizes After Register") do
-          newly_registered_userfiles.each do |userfile|
-            userfile.set_size! rescue true
+    flash[:notice] ||= ''
+
+    # Unregister the given userfiles in background.
+    userfiles = userfiles_from_basenames(@provider, @as_user, params[:basenames])
+    succeeded, failed = [], {}
+    erasing = params.has_key?(:delete)
+
+    CBRAIN.spawn_with_active_records_if(
+      [:html, :js].include?(request.format.to_sym),
+      current_user,
+      "Unregister files (data_provider: #{@provider.id})"
+    ) do
+      userfiles.reject { |b,u| u.blank? }.shuffle.each do |basename, userfile|
+        begin
```

Comment:
Just before the begin block, insert a `$0 = ""` counter line so that the admin can monitor the progress when doing 'ps'. See above in 'register'.